### ISSUE TITLE
CheckRepoStatus fails for missing database

### ISSUE THREAD

**Issue by LostKobrakai** [phoenixframework/phoenix_ecto#176]
On phoenix_ecto 4.6.0 I get the following logs on a freshly generated phoenix application:

```
[error] Postgrex.Protocol (#PID<0.273.0>) failed to connect: ** (Postgrex.Error) FATAL 3D000 (invalid_catalog_name) database "planner_dev" does not exist
…
```

The plug should afaik catch that state, but seems to fail trying to run migrations:

# DBConnection.ConnectionError at GET /

Exception:

    ** (DBConnection.ConnectionError) connection not available and request was dropped from queue after 2769ms. This means requests are coming in and your connection pool cannot serve them fast enough. You can address this by:
    
      1. Ensuring your database is available and that you can connect to it
      2. Tracking down slow queries and making sure they are running fast enough
      3. Increasing the pool_size (although this increases resource consumption)
      4. Allowing requests to wait longer by increasing :queue_target and :queue_interval
    
    See DBConnection.start_link/2 for more information
    
        (ecto_sql 3.11.1) lib/ecto/adapters/sql.ex:1054: Ecto.Adapters.SQL.raise_sql_call_error/1
        (elixir 1.16.2) lib/enum.ex:1700: Enum."-map/2-lists^map/1-1-"/2
        (ecto_sql 3.11.1) lib/ecto/adapters/sql.ex:1161: Ecto.Adapters.SQL.execute_ddl/4
        (ecto_sql 3.11.1) lib/ecto/migrator.ex:755: Ecto.Migrator.verbose_schema_migration/3
        (ecto_sql 3.11.1) lib/ecto/migrator.ex:563: Ecto.Migrator.lock_for_migrations/4
        (ecto_sql 3.11.1) lib/ecto/migrator.ex:485: Ecto.Migrator.migrations/3
        (phoenix_ecto 4.6.0) lib/phoenix_ecto/check_repo_status.ex:99: Phoenix.Ecto.CheckRepoStatus.migrations/3
        (phoenix_ecto 4.6.0) lib/phoenix_ecto/check_repo_status.ex:57: Phoenix.Ecto.CheckRepoStatus.check_pending_migrations!/2
        (phoenix_ecto 4.6.0) lib/phoenix_ecto/check_repo_status.ex:31: anonymous fn/3 in Phoenix.Ecto.CheckRepoStatus.call/2
        (elixir 1.16.2) lib/enum.ex:2528: Enum."-reduce/3-lists^foldl/2-0-"/3
        (phoenix_ecto 4.6.0) lib/phoenix_ecto/check_repo_status.ex:30: Phoenix.Ecto.CheckRepoStatus.call/2
        (planner 0.1.0) lib/planner_web/endpoint.ex:1: PlannerWeb.Endpoint.plug_builder_call/2
        (planner 0.1.0) deps/plug/lib/plug/debugger.ex:136: PlannerWeb.Endpoint."call (overridable 3)"/2
        (planner 0.1.0) lib/planner_web/endpoint.ex:1: PlannerWeb.Endpoint.call/2
        (phoenix 1.7.12) lib/phoenix/endpoint/sync_code_reload_plug.ex:22: Phoenix.Endpoint.SyncCodeReloadPlug.do_call/4
        (bandit 1.5.2) lib/bandit/pipeline.ex:124: Bandit.Pipeline.call_plug!/2
        (bandit 1.5.2) lib/bandit/pipeline.ex:36: Bandit.Pipeline.run/4
        (bandit 1.5.2) lib/bandit/http1/handler.ex:12: Bandit.HTTP1.Handler.handle_data/3
        (bandit 1.5.2) lib/bandit/delegating_handler.ex:18: Bandit.DelegatingHandler.handle_data/3
        (bandit 1.5.2) /Users/benni/Work/ambihome/Projekte/research/05_elixir/planner/deps/thousand_island/lib/thousand_island/handler.ex:411: Bandit.DelegatingHandler.handle_continue/2
    

Code:

`lib/ecto/adapters/sql.ex`

    1049     defp raise_sql_call_error(%DBConnection.OwnershipError{} = err) do
    1050       message = err.message &lt;&gt; "\nSee Ecto.Adapters.SQL.Sandbox docs for more information."
    1051       raise %{err | message: message}
    1052     end
    1053   
    1054&gt;    defp raise_sql_call_error(err), do: raise(err)
    1055   
    1056     @doc false
    1057     def reduce(adapter_meta, statement, params, opts, acc, fun) do
    1058       %{pid: pool, telemetry: telemetry, sql: sql, opts: default_opts} = adapter_meta
    1059       opts = with_log(telemetry, params, opts ++ default_opts)
    
`lib/enum.ex`

    No code available.

`lib/ecto/adapters/sql.ex`

    1156     def execute_ddl(meta, conn, definition, opts) do
    1157       ddl_logs =
    1158         definition
    1159         |&gt; conn.execute_ddl()
    1160         |&gt; List.wrap()
    1161&gt;        |&gt; Enum.map(&amp;query!(meta, &amp;1, [], opts))
    1162         |&gt; Enum.flat_map(&amp;conn.ddl_logs/1)
    1163   
    1164       {:ok, ddl_logs}
    1165     end
    1166   
    
`lib/ecto/migrator.ex`

    750       end)
    751     end
    752   
    753     defp verbose_schema_migration(repo, reason, fun) do
    754       try do
    755&gt;        fun.()
    756       rescue
    757         error -&gt;
    758           Logger.error("""
    759           Could not #{reason}. This error usually happens due to the following:
    760   
    
`lib/ecto/migrator.ex`

    558   
    559       try do
    560         config = repo.config()
    561   
    562         unless skip_table_creation do
    563&gt;          verbose_schema_migration(repo, "create schema migrations table", fn -&gt;
    564             SchemaMigration.ensure_schema_migrations_table!(repo, config, opts)
    565           end)
    566         end
    567   
    568         {migration_repo, query, all_opts} = SchemaMigration.versions(repo, config, opts[:prefix])
    
`lib/ecto/migrator.ex`

    480             [{:up | :down, id :: integer(), name :: String.t()}]
    481     def migrations(repo, directories, opts \\ []) do
    482       directories = List.wrap(directories)
    483   
    484       repo
    485&gt;      |&gt; migrated_versions(opts)
    486       |&gt; collect_migrations(directories)
    487       |&gt; Enum.sort_by(fn {_, version, _} -&gt; version end)
    488     end
    489   
    490     use GenServer
    
`lib/phoenix_ecto/check_repo_status.ex`

    94           migrations = get_migrations(migration_fn, repo, migration_directories, migration_opts)
    95           {:ok, migrations}
    96   
    97         :error -&gt;
    98           if Code.ensure_loaded?(Ecto.Migrator) do
    99&gt;            {:ok, Ecto.Migrator.migrations(repo, migration_directories, migration_opts)}
    100           else
    101             :error
    102           end
    103       end
    104     end
    
`lib/phoenix_ecto/check_repo_status.ex`

    52     end
    53   
    54     defp check_pending_migrations!(repo, opts) do
    55       repo_status =
    56         with {:ok, migration_directories} &lt;- migration_directories(repo, opts),
    57&gt;             {:ok, migrations} &lt;- migrations(repo, migration_directories, opts) do
    58           {:ok, migration_directories, migrations}
    59         end
    60   
    61       case repo_status do
    62         {:ok, migration_directories, migrations} -&gt;
    
`lib/phoenix_ecto/check_repo_status.ex`

    26   
    27     def call(%Conn{} = conn, opts) do
    28       repos = Application.get_env(opts[:otp_app], :ecto_repos, [])
    29   
    30       for repo &lt;- repos, Process.whereis(repo) do
    31&gt;        unless check_pending_migrations!(repo, opts) do
    32           check_storage_up!(repo)
    33         end
    34       end
    35   
    36       conn
    
`lib/enum.ex`

    No code available.

`lib/phoenix_ecto/check_repo_status.ex`

    25     end
    26   
    27     def call(%Conn{} = conn, opts) do
    28       repos = Application.get_env(opts[:otp_app], :ecto_repos, [])
    29   
    30&gt;      for repo &lt;- repos, Process.whereis(repo) do
    31         unless check_pending_migrations!(repo, opts) do
    32           check_storage_up!(repo)
    33         end
    34       end
    35   
    
`lib/planner_web/endpoint.ex`

    1&gt;  defmodule PlannerWeb.Endpoint do
    2     use Phoenix.Endpoint, otp_app: :planner
    3   
    4     # The session will be stored in the cookie and signed,
    5     # this means its contents can be read but not tampered with.
    6     # Set :encryption_salt if you would also like to encrypt it.
    
`deps/plug/lib/plug/debugger.ex`

    131             case conn do
    132               %Plug.Conn{path_info: ["__plug__", "debugger", "action"], method: "POST"} -&gt;
    133                 Plug.Debugger.run_action(conn)
    134   
    135               %Plug.Conn{} -&gt;
    136&gt;                super(conn, opts)
    137             end
    138           rescue
    139             e in Plug.Conn.WrapperError -&gt;
    140               %{conn: conn, kind: kind, reason: reason, stack: stack} = e
    141               Plug.Debugger.__catch__(conn, kind, reason, stack, @plug_debugger)
    
`lib/planner_web/endpoint.ex`

    1&gt;  defmodule PlannerWeb.Endpoint do
    2     use Phoenix.Endpoint, otp_app: :planner
    3   
    4     # The session will be stored in the cookie and signed,
    5     # this means its contents can be read but not tampered with.
    6     # Set :encryption_salt if you would also like to encrypt it.
    
`lib/phoenix/endpoint/sync_code_reload_plug.ex`

    17   
    18     def call(conn, {endpoint, opts}), do: do_call(conn, endpoint, opts, true)
    19   
    20     defp do_call(conn, endpoint, opts, retry?) do
    21       try do
    22&gt;        endpoint.call(conn, opts)
    23       rescue
    24         exception in [UndefinedFunctionError] -&gt;
    25           case exception do
    26             %UndefinedFunctionError{module: ^endpoint} when retry? -&gt;
    27               # Sync with the code reloader and retry once
    
`lib/bandit/pipeline.ex`

    119       end
    120     end
    121   
    122     @spec call_plug!(Plug.Conn.t(), plug_def()) :: Plug.Conn.t() | no_return()
    123     defp call_plug!(%Plug.Conn{} = conn, {plug, plug_opts}) when is_atom(plug) do
    124&gt;      case plug.call(conn, plug_opts) do
    125         %Plug.Conn{} = conn -&gt; conn
    126         other -&gt; raise("Expected #{plug}.call/2 to return %Plug.Conn{} but got: #{inspect(other)}")
    127       end
    128     end
    129   
    
`lib/bandit/pipeline.ex`

    31         conn = build_conn!(transport, method, request_target, headers, opts)
    32         span = Bandit.Telemetry.start_span(:request, measurements, Map.put(metadata, :conn, conn))
    33   
    34         try do
    35           conn
    36&gt;          |&gt; call_plug!(plug)
    37           |&gt; maybe_upgrade!()
    38           |&gt; case do
    39             {:no_upgrade, conn} -&gt;
    40               %Plug.Conn{adapter: {_mod, adapter}} = conn = commit_response!(conn)
    41               Bandit.Telemetry.stop_span(span, adapter.metrics, %{conn: conn})
    
`lib/bandit/http1/handler.ex`

    7     @impl ThousandIsland.Handler
    8     def handle_data(data, socket, state) do
    9       transport = %Bandit.HTTP1.Socket{socket: socket, buffer: data, opts: state.opts}
    10       connection_span = ThousandIsland.Socket.telemetry_span(socket)
    11   
    12&gt;      case Bandit.Pipeline.run(transport, state.plug, connection_span, state.opts) do
    13         {:ok, transport} -&gt; maybe_keepalive(transport, state)
    14         {:error, _reason} -&gt; {:close, state}
    15         {:upgrade, _transport, :websocket, opts} -&gt; do_websocket_upgrade(opts, state)
    16       end
    17     end
    
`lib/bandit/delegating_handler.ex`

    13       |&gt; handle_bandit_continuation(socket)
    14     end
    15   
    16     @impl ThousandIsland.Handler
    17     def handle_data(data, socket, %{handler_module: handler_module} = state) do
    18&gt;      handler_module.handle_data(data, socket, state)
    19       |&gt; handle_bandit_continuation(socket)
    20     end
    21   
    22     @impl ThousandIsland.Handler
    23     def handle_shutdown(socket, %{handler_module: handler_module} = state) do
    
`/Users/benni/Work/ambihome/Projekte/research/05_elixir/planner/deps/thousand_island/lib/thousand_island/handler.ex`

    406         # to state in case the `c:handle_connection/2` callback raises an error.
    407         # This ensures that the `c:terminate/2` calls below are able to properly
    408         # close down the process
    409         @impl GenServer
    410         def handle_continue(:handle_connection, {%ThousandIsland.Socket{} = socket, state}) do
    411&gt;          __MODULE__.handle_connection(socket, state)
    412           |&gt; handle_continuation(socket)
    413         end
    414   
    415         # Called if the remote end closed the connection before we could initialize it
    416         @impl GenServer
    

### Params

    %{}

### Request info

  * URI: http://localhost:4000/
  * Query string: 

### Headers
  
  * accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8
  * accept-encoding: gzip, deflate, br, zstd
  * accept-language: de,en-US;q=0.7,en;q=0.3
  * connection: keep-alive
  * cookie: user_data=eyJoZXhfY29sb3IiOiIjOUVEOUNDIiwibmFtZSI6bnVsbH0=; lb:user_data=eyJoZXhfY29sb3IiOiIjRjQ3MkI2IiwibmFtZSI6bnVsbH0=; locale=en; _connect_key=SFMyNTY.g3QAAAABbQAAAAtfY3NyZl90b2tlbm0AAAAYZk5PNUxPRy03T1JkWF9zNEFLQmFGVmNx.7Pt_ufa1myhRb2wVFV5ISvwArqerNZzgOUmZ25bAPcQ; _social_key=SFMyNTY.g3QAAAABbQAAAAtfY3NyZl90b2tlbm0AAAAYWHE2OVFJWGR1bjFoeGppNGJwbUEyOHRl.oyubdZBz-Qf-B-qrJLrlN-aRQ5Duq_wwHLbUOBOWk3I; _playground_key=SFMyNTY.g3QAAAABbQAAAAtfY3NyZl90b2tlbm0AAAAYbVJLYXAySmo3QjZQOThQNVEzSUt0Vkx2.yUsAgH6_bY_iDLjbF9BeA_DeMM31PLd8fn3k67nYu0A; _overlay_key=SFMyNTY.g3QAAAABbQAAAAtfY3NyZl90b2tlbm0AAAAYb2xZLWtwWWpGYUJvOUlxWHpYaDVsRUxq.5mdkCHjuteUA6c7Us8Nhi4ZKhaJoCYrqjFa3_yASp0Y; lb:session=SFMyNTY.g3QAAAAFbQAAAAs1NDM1MDp0b2tlbm0AAAAgyYY2e_1DBF6eWdlhLt7FukCy-Tx_zgxFZZ6g8LYkhodtAAAAC19jc3JmX3Rva2VubQAAABhhZGttZ0l3Wm0xQUtXQk1WRjBjTTBwcUdtAAAAD2N1cnJlbnRfdXNlcl9pZG0AAAAgaDNjNDZya2w3NDVsM3B1eTZvcm0ybXE0Z290dTZpeG1tAAAADWlkZW50aXR5X2RhdGF0AAAAAXcCaWRtAAAAIGgzYzQ2cmtsNzQ1bDNwdXk2b3JtMm1xNGdvdHU2aXhtbQAAAAl1c2VyX2RhdGF0AAAAAm0AAAAJaGV4X2NvbG9ybQAAAAcjRjQ3MkI2bQAAAARuYW1ldwNuaWw.aJ-PG2GVOWrmdmj0d0QY8jp0I47WFMw9qjiqNbP8dbk; lb:probe_cookie=; _shared_key=SFMyNTY.g3QAAAABbQAAAAtfY3NyZl90b2tlbm0AAAAYcWRxbUVXYi0yay1zMVMxa04wVUZxVW1D.gTD42A3f8DyPBI2xcJwvjUSHkGgFZebhsy5sdZHCukc
  * dnt: 1
  * host: localhost:4000
  * priority: u=1
  * sec-fetch-dest: document
  * sec-fetch-mode: navigate
  * sec-fetch-site: none
  * sec-fetch-user: ?1
  * upgrade-insecure-requests: 1
  * user-agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:127.0) Gecko/20100101 Firefox/127.0

### Session

    nil

**Comment by josevalim** [comment]
I assume this is happening for quite some time, because I went back in time and I could not trivially find a working version, even going back several SQL versions.

**Comment by josevalim** [comment]
Broken in #156.

**Comment by josevalim** [comment]
Fixed in bf499065a43cbde3078a04a2050a8f2d5c4fff9e.
